src/main/java/com/google/sps/servlets/utils/Statistics.java

```diff
@@ -14,29 +14,90 @@
+
+  private static final BigDecimal UPPER_END = BigDecimal.valueOf(1.0);
+
+  private static final BigDecimal LOWER_END = BigDecimal.valueOf(-1.0);
+
+  // Contains sentiment scores in the range [-1, 1] with given intervals.
   private Map<Range, Integer> aggregateValues;
   private double averageScore;
 
+  public Map<Range, Integer> getAggregateValues() {
+    return aggregateValues;
+  }
+
+  public double getAverageScore() {
+    return averageScore;
+  }
+
+  /**
+   * Constructor of Statistics to filter out invalid sentiment scores, set aggregate hash map and
+   * average score
+   *
+   * @param sentimentScores given score values
+   */
   public Statistics(List<Double> sentimentScores) {
+    sentimentScores =
+        sentimentScores.stream()
+            .filter(score -> (score >= LOWER_END_VAL && score <= UPPER_END_VAL))
+            .collect(Collectors.toList());
     setAggregateScores(sentimentScores);
     setAverageScore(sentimentScores);
   }
 
+  /**
+   * Categorize all score values into different range intervals and count the frequency for each
+   * interval, and set the aggregatedValues.
+   *
+   * @param sentimentScores a list of score values from -1.0 to 1.0
+   */
   private void setAggregateScores(List<Double> sentimentScores) {
-    // TODO(Xin): Add sorting code
     aggregateValues = new HashMap<>();
-    aggregateValues.put(new Range(-1.0, 1), 1);
+
+    // Initialize the HashMap with intervals
+    for (BigDecimal tempPoint = LOWER_END;
+        tempPoint.compareTo(UPPER_END) < 0;
+        tempPoint = tempPoint.add(INTERVAL)) {
+      Range currentRange = new Range(tempPoint, UPPER_END.min(tempPoint.add(INTERVAL)));
+      aggregateValues.put(currentRange, 0);
+    }
+    // Add score's interval to different ranges two sorting with and two pointers pop-up
+    sentimentScores.sort(Comparator.naturalOrder());
+    BigDecimal curPoint = LOWER_END;
+    int scoreIdx = 0;
+    while ((scoreIdx < sentimentScores.size()) && (curPoint.compareTo(UPPER_END) < 0)) {
+      // check available interval for scoreVal
+      BigDecimal nextPoint = curPoint.add(INTERVAL);
+      BigDecimal scorePoint = BigDecimal.valueOf(sentimentScores.get(scoreIdx));
+      if (scorePoint.compareTo(nextPoint) < 0 || (nextPoint.compareTo(UPPER_END) == 0)) {
+        Range foundRange = new Range(curPoint, nextPoint);
+        aggregateValues.put(foundRange, aggregateValues.get(foundRange) + 1);
+        scoreIdx += 1;
+      } else {
+        curPoint = nextPoint;
+      }
+    }
   }
 
+  /**
+   * Set the average score of given sentiment scores. Returns -99 if the average score is not valid
+   * or none of the sentiment scores is valid.
+   *
+   * @param sentimentScores a list of score values from -1.0 to 1.0
+   */
   private void setAverageScore(List<Double> sentimentScores) {
-    // TODO(Xin): Add average score method
-    averageScore = 0;
+    averageScore = sentimentScores.stream().mapToDouble(i -> i).average().orElse(-99);
```

Comment:
Let's change this to thrown an exception instead of returning this "magic number" (https://en.wikipedia.org/wiki/Magic_number_(programming)).

If we end up not having any sentiment scores, then we have bigger problems than not being able to calculate an average.

This could be changed to:

`averageScore = sentimentScores.stream().mapToDouble(i -> i).average().orElseThrow(() -> new RuntimeException("Unable to calculate sentiment average due to empty input list.");`